pyteal/ast/subroutine.py

```
@@ -446,53 +620,93 @@ def evaluateSubroutine(subroutine: SubroutineDefinition) -> SubroutineDeclaratio
+        Type 4. (ABI-output-arg) generates a new instance of the ABI value,
+            and appends a return expression of stored value of the ABI keyword value.
     """
 
     def var_n_loaded(
         param: str,
-    ) -> Tuple[ScratchVar, Union[ScratchVar, abi.BaseType, Expr]]:
-        loaded: Union[ScratchVar, abi.BaseType, Expr]
-        argVar: ScratchVar
+    ) -> tuple[ScratchVar, ScratchVar | abi.BaseType | Expr]:
+        _loaded_var: ScratchVar | abi.BaseType | Expr
+        _argument_var: ScratchVar
 
         if param in subroutine.by_ref_args:
-            argVar = DynamicScratchVar(TealType.anytype)
-            loaded = argVar
+            _argument_var = DynamicScratchVar(TealType.anytype)
+            _loaded_var = _argument_var
         elif param in subroutine.abi_args:
             internal_abi_var = subroutine.abi_args[param].new_instance()
-            argVar = internal_abi_var.stored_value
-            loaded = internal_abi_var
+            _argument_var = internal_abi_var.stored_value
+            _loaded_var = internal_abi_var
         else:
-            argVar = ScratchVar(TealType.anytype)
-            loaded = argVar.load()
+            _argument_var = ScratchVar(TealType.anytype)
+            _loaded_var = _argument_var.load()
 
-        return argVar, loaded
+        return _argument_var, _loaded_var
 
     args = subroutine.arguments()
-    argumentVars, loadedArgs = zip(*map(var_n_loaded, args)) if args else ([], [])
+    args = [arg for arg in args if arg not in subroutine.output_kwarg]
+
+    argument_vars: list[ScratchVar] = []
+    loaded_args: list[ScratchVar | Expr | abi.BaseType] = []
+    for arg in args:
+        argument_var, loaded_arg = var_n_loaded(arg)
+        argument_vars.append(argument_var)
+        loaded_args.append(loaded_arg)
+
+    abi_output_kwargs: dict[str, abi.BaseType] = {}
+    if len(subroutine.output_kwarg) > 1:
```

I recommend putting this validation at the top so you can "fail fast"

---

good point